File: tests/sabotage/compare/nochange.js
/* No-change comparison: master's index.html vs this commit's, for every
   current catalog card. Compares computeInference() field by field on a full
   grid, and every renderer's HTML, the copied Markdown report, the vLLM
   command and the board advice on a strided sample of it. Also runs each
   version's own readInputState() through a DOM stub, card by card.

   Run: node tmp/nochange.js [stride]   (stride defaults to 13) */
const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');
const os = require('os');
const ROOT = execFileSync('git', ['rev-parse', '--show-toplevel'],
  { cwd: __dirname, encoding: 'utf8' }).trim();
const BASE_REF = process.argv[2] || 'master';

/* Both sides come from git rather than checked-in copies: a snapshot of a 200 KB
   engine goes stale the moment either side moves, and a comparison against a stale
   "master" reports zero differences for the wrong reason. */
function atRef(ref, rel) {
  const blob = execFileSync('git', ['show', `${ref}:${rel}`],
    { cwd: ROOT, maxBuffer: 256 * 1024 * 1024 });
  const tmp = path.join(os.tmpdir(), `nochange_${ref.replace(/\W/g, '_')}_${path.basename(rel)}`);
  fs.writeFileSync(tmp, blob);
  return tmp;
}
const STRIDE = Number(process.argv[2] || 13);

function load(file) {
  const html = fs.readFileSync(file, 'utf8');
  const out = {};
  const values = { 'train-method': 'lora', 'train-optimizer': 'adam', 'train-grad-ckpt': '1',
    'train-batch-size': '1', 'train-seq-len': '2048', 'train-lora-rank': '16',
    'train-lora-targets': '4', 'train-hidden-size': '4096' };
  const el = (id) => ({
    get innerHTML() { return out[id] || ''; }, set innerHTML(v) { out[id] = v; },
    get value() { return values[id] !== undefined ? values[id] : ''; }, set value(v) { values[id] = v; },
    style: {}, options: [], selectedOptions: [{ dataset: { q: '' } }], checked: true,
    classList: { add() {}, remove() {} },
  });
  const cache = {};
  const document = {
    getElementById: (id) => (cache[id] = cache[id] || el(id)),
    body: { classList: { add() {}, remove() {}, contains: () => true } },
    querySelector: () => ({ textContent: '', innerHTML: '', parentElement: null }),
    querySelectorAll: () => [],
  };
  const start = html.indexOf('function getVal(id)');
  const before = [/^const GPU_TABLE = \{[\s\S]*?\n\};$/m, /^const BENCHMARK_DATA = \{[\s\S]*?\n\};$/m,
                  /^const MODEL_PRESETS = \{[\s\S]*?\n\};$/m, /^const WORKLOAD_PROFILES = \{[\s\S]*?\n\};$/m]
    .map(re => html.match(re)[0]).filter(d => html.indexOf(d) < start).join('\n');
  const navigator = { clipboard: { writeText: () => Promise.resolve() } };
  const src = html.slice(start, html.indexOf('function updateURLHash()'));
  const names = [...src.matchAll(/^function (\w+)\(/gm)].map(m => m[1]);
  const api = new Function('document', 'navigator', `
    let currentAttn = { mode: 'standard', window: 0, localLayers: 0, mlaDim: 0 };
    let currentModelMaxCtx = 131072, importedModelId = null, urlRestoreLost = [];
    let savedSnapshots = [];
    ${before}
    ${src}
    return { ${names.join(', ')}, GPU_TABLE, MODEL_PRESETS,
             pushSnapshot: (s, c) => savedSnapshots.push({ state: s, computed: c, name: 'snap' }),
             clearSnapshots: () => { savedSnapshots.length = 0; } };`)(document, navigator);
  const params = {};
  for (const m of src.matchAll(/function (render\w+)\(([^)]*)\)/g))
    params[m[1]] = m[2].split(',').map(x => x.trim().split(/[=\s]/)[0]).filter(Boolean);
  return { ...api, out, cache, values, html, params, renderers: names.filter(n => /^render/.test(n)) };
}

const NEW = load(path.join(ROOT, 'index.html'));       // the working tree
const OLD = load(atRef(BASE_REF, 'index.html'));

// ---- readInputState through a DOM stub, each version's own -------------------
const domStub = (gpuKey, interconnect, prec = '2', q = '') => {
  const el = (v, extra = {}) => ({ value: v, style: {}, options: [], ...extra });
  const fields = {
    'gpu-model': el(gpuKey), 'interconnect': el(interconnect),
    'param-count': el('8'), 'active-percent': el('100'), 'layer-count': el('32'),
    'kv-head-count': el('8'), 'head-dim': el('128'), 'shared-experts': el('0'),
    'context-length': el('8192'), 'concurrency': el('1'), 'gpu-count': el('2'),
    'shared-prefix': el('0'), 'prefix-caching': el('1'),
    'kv-precision': el('2'), 'preset': el(''),
    'weight-precision': el(prec, { selectedOptions: [{ dataset: { q } }] }),
  };
  return { getElementById: (id) => fields[id] };
};
const readInputStateFor = (v, gpuKey, interconnect, prec, q) => {
  const stop = v.html.indexOf('\nconst PERF = {');
  const src = v.html.slice(v.html.indexOf('function getVal(id)'), stop);
  const fn = new Function('document', 'GPU_TABLE', `
    let currentAttn = { mode: 'standard', window: 0, localLayers: 0, mlaDim: 0 };
    let currentModelMaxCtx = 131072, importedModelId = null;
    ${src}
    return readInputState;`)(domStub(gpuKey, interconnect, prec, q), v.GPU_TABLE);
  return fn();
};
let stateDiffs = 0, stateChecked = 0;
for (const key of Object.keys(NEW.GPU_TABLE)) {
  for (const ic of ['1', '0']) for (const [prec, q] of [['2', ''], ['1', 'fp8'], ['0.5', 'awq'], ['0.63', 'gguf']]) {
    const a = readInputStateFor(OLD, key, ic, prec, q), b = readInputStateFor(NEW, key, ic, prec, q);
    stateChecked++;
    const keys = new Set([...Object.keys(a), ...Object.keys(b)]);
    for (const k of keys) {
      if (!(k in a)) { stateDiffs++; console.log(`readInputState ${key}: new adds field ${k}`); continue; }
      if (!(k in b)) { stateDiffs++; console.log(`readInputState ${key}: new dropped field ${k}`); continue; }
      if (JSON.stringify(a[k]) !== JSON.stringify(b[k])) { stateDiffs++; console.log(`readInputState ${key}/${ic}/${prec}: ${k} ${JSON.stringify(a[k])} -> ${JSON.stringify(b[k])}`); }
    }
  }
}
console.log(`readInputState: ${stateChecked} states compared, ${stateDiffs} differences`);

// ---- the grid ----------------------------------------------------------------
const P = NEW.MODEL_PRESETS;
const MODELS = ['llama31-8b', 'llama31-70b', 'qwen3-30b', 'gemma4-26b', 'dsv3-671b', 'mistral-lg-123b'];
const COUNTS = [1, 2, 4, 8, 9, 12, 16, 64, 128];
const PRECS = [[2, ''], [1, 'fp8'], [0.5, 'awq'], [0.63, 'gguf']];
const KVS = [2, 1];
const CTX = [[4096, 1], [8192, 20], [32768, 64], [131072, 1]];
const NV = [true, false];
const PREFIX = [[0, true], [2048, true], [2048, false]];

function stateFor(key, g, count, model, [bpp, q], kv, [ctx, conc], nv, [pre, pc]) {
  const p = P[model];
  return {
    params: p.p, activePercent: p.a, bytesPerParam: bpp, quantMethod: q, layers: p.l, kvHeads: p.kv,
    headDim: p.hd, sharedExperts: p.se, contextLength: ctx, concurrency: conc, sharedPrefix: pre,
    prefixCaching: pc, gpuCount: count, gpuGB: g.gb, gpuBandwidth: g.bw, gpuTFLOPS: g.tflops,
    gpuHyperCost: g.hyper, gpuSpecCost: g.spec, gpuSpotCost: g.spot, gpuName: g.name.replace(/ GB$/, 'GB'),
    gpuDevices: g.devices, gpuKey: key, perfKey: g.perfKey, vendor: g.vendor,
    gpuFp8: !!(g.caps && g.caps.fp8), hasNVLink: g.form === 'sxm' && nv, kvBytesPerValue: kv,
    presetKey: '', hfModelId: null, attnMode: p.attn || 'standard', swaWindow: p.swaWin || 0,
    swaLocalLayers: Math.min(p.swaLocal || 0, p.l), mlaLatentDim: p.mlaDim || 0, modelMaxCtx: p.maxCtx || 131072,
  };
}
const same = (a, b) => (Number.isNaN(a) && Number.isNaN(b)) || JSON.stringify(a) === JSON.stringify(b);

let n = 0, computeDiffs = 0, renderDiffs = 0, rendered = 0, newOnly = new Set(), oldOnly = new Set();
const diffSamples = [];
const renderAll = (v, st) => {
  for (const k of Object.keys(v.out)) delete v.out[k];
  v.clearSnapshots();
  const c = v.computeInference(st);
  v.pushSnapshot(st, c);
  for (const name of v.renderers) {
    const args = (v.params[name] || []).map(pn => (pn === 'computed' ? c : pn === 'state' ? st : undefined));
    v[name](...args);
  }
  const res = { ...v.out };
  res['(copied report)'] = v.exportSummary(st, c);
  res['(command)'] = v.buildVllmCommand(st, c, 'm');
  if (!c.fits) res['(advice)'] = v.boardsAdvice(st, c);
  return res;
};
for (const [key, g] of Object.entries(NEW.GPU_TABLE))
  for (const count of COUNTS) for (const model of MODELS) for (const prec of PRECS) for (const kv of KVS)
    for (const ctx of CTX) for (const nv of NV) for (const pre of PREFIX) {
      if (!nv && g.form !== 'sxm') continue; // hasNVLink is false either way; skip the duplicate
      const st = stateFor(key, g, count, model, prec, kv, ctx, nv, pre);
      const a = OLD.computeInference(st), b = NEW.computeInference(st);
      n++;
      for (const k of Object.keys(a)) {
        if (!(k in b)) { oldOnly.add(k); continue; }
        if (!same(a[k], b[k])) { computeDiffs++; if (diffSamples.length < 10) diffSamples.push(`${key} x${count} ${model} ${prec} kv${kv} ${ctx} nv${nv} ${pre}: ${k} ${JSON.stringify(a[k])} -> ${JSON.stringify(b[k])}`); }
      }
      for (const k of Object.keys(b)) if (!(k in a)) newOnly.add(k);
      if (b.throughputModelled !== true) { computeDiffs++; diffSamples.push(`${key}: throughputModelled=${b.throughputModelled}`); }
      if (n % STRIDE === 0) {
        rendered++;
        const ra = renderAll(OLD, st), rb = renderAll(NEW, st);
        for (const id of new Set([...Object.keys(ra), ...Object.keys(rb)])) {
          if (ra[id] !== rb[id]) {
            renderDiffs++;
            if (diffSamples.length < 10) {
              const i = [...(ra[id] || '')].findIndex((ch, idx) => ch !== (rb[id] || '')[idx]);
              diffSamples.push(`${key} x${count} ${model} ${prec} kv${kv} ${ctx} nv${nv} ${pre}: ${id} differs at ${i}: old "${(ra[id] || '').slice(Math.max(0, i - 60), i + 60)}" new "${(rb[id] || '').slice(Math.max(0, i - 60), i + 60)}"`);
            }
          }
        }
      }
    }
console.log(`computeInference: ${n} configurations, ${computeDiffs} field differences`);
console.log(`  fields only in new: [${[...newOnly].join(', ')}]  only in master: [${[...oldOnly].join(', ')}]`);
console.log(`renderers + copied report + command + advice: ${rendered} configurations, ${renderDiffs} differing surfaces`);
for (const d of diffSamples) console.log('  ' + d);

File: tests/sabotage/compare/nochange.py
#!/usr/bin/env python3
"""No-change comparison: master's generate_report.py vs this commit's, for every
current catalog card. compute() on a full grid through each version's own
from_cli_args(); every story string of the PDF on a strided sample; the real
PDF bytes (timestamps pinned, reportlab invariant mode) on a smaller sample;
from_json and interactive_mode builders on a few.

Both versions come from git, not from checked-in copies. Snapshots of a 70 KB
engine go stale the moment either side moves, and a comparison against a stale
"master" reports zero differences for the wrong reason.

Run: python3 tests/sabotage/compare/nochange.py [base-ref] [stride] [pdf_stride]
     base-ref defaults to master; the other side is always the working tree.

Every field is compared, with no exemptions. An earlier version excused `perfKey`
because the commit it was written for was the one adding it — which made the
comparison asymmetric, reported five false differences once both sides had the
field, and, worse, would have hidden a real perfKey change in any later commit.
A field this harness is told to ignore is a field it cannot report on, so when a
change legitimately adds one, read the diff it prints rather than silencing it.
"""
import contextlib, importlib.util, io, itertools, os, subprocess, sys, tempfile, types, unittest.mock, datetime as _dt
ROOT = subprocess.check_output(["git", "rev-parse", "--show-toplevel"],
                               cwd=os.path.dirname(os.path.abspath(__file__)),
                               text=True).strip()
BASE_REF = sys.argv[1] if len(sys.argv) > 1 else "master"
STRIDE = int(sys.argv[2]) if len(sys.argv) > 2 else 29
PDF_STRIDE = int(sys.argv[3]) if len(sys.argv) > 3 else 977


def at_ref(ref, rel):
    """The file as it stands at <ref>, written somewhere importable."""
    blob = subprocess.check_output(["git", "show", f"{ref}:{rel}"], cwd=ROOT)
    fd, tmp = tempfile.mkstemp(suffix="_" + os.path.basename(rel))
    os.write(fd, blob)
    os.close(fd)
    return tmp

import reportlab.rl_config
reportlab.rl_config.invariant = 1


def load(name, rel):
    spec = importlib.util.spec_from_file_location(name, os.path.join(ROOT, rel))
    m = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(m)
    return m


NEW = load("gr_new", "generate_report.py")                       # the working tree
OLD = load("gr_old", os.path.relpath(at_ref(BASE_REF, "generate_report.py"), ROOT))


class FixedDT(_dt.datetime):
    @classmethod
    def now(cls, tz=None):
        return cls(2026, 9, 17, 12, 0, 0)


for m in (NEW, OLD):
    m.datetime = FixedDT

MODELS = ["llama31-8b", "llama31-70b", "qwen3-30b", "gemma4-26b", "dsv3-671b", "mistral-lg-123b"]
COUNTS = [1, 2, 4, 8, 9, 12, 16, 64, 128]
PRECS = ["bf16", "fp8", "awq", "q4km"]
KVS = [False, True]
CTX = [(4096, 1), (8192, 20), (32768, 64), (131072, 1)]
NV = [False, True]  # no_nvlink


def args_for(gpu, count, model, prec, fp8kv, ctx, conc, no_nv):
    return types.SimpleNamespace(preset=model, gpu=gpu, prec=prec, ctx=ctx, conc=conc,
                                 ngpu=count, no_nvlink=no_nv, fp8_kv=fp8kv)


def story_strings(m, cfg):
    obj = m.ReportCard(cfg, output_path=os.devnull)
    seen = []

    def harvest(item):
        text = getattr(item, "text", None)
        if text:
            seen.append(text)
        for line in getattr(item, "lines", None) or []:
            seen.append(line if isinstance(line, str) else str(line))
        for row in getattr(item, "_cellvalues", []):
            for cell in row:
                harvest(cell) if hasattr(cell, "text") or hasattr(cell, "contents") else seen.append(str(cell))
        for child in getattr(item, "contents", []) or []:
            harvest(child)

    real_build = m.SimpleDocTemplate.build
    captured = {}
    try:
        m.SimpleDocTemplate.build = lambda self, story, **kw: captured.__setitem__("story", story)
        obj.generate()
        for item in captured.get("story", []):
            harvest(item)
    finally:
        m.SimpleDocTemplate.build = real_build
    return seen


def pdf_bytes(m, cfg):
    import tempfile
    fd, out = tempfile.mkstemp(suffix=".pdf")
    os.close(fd)
    try:
        with contextlib.redirect_stdout(io.StringIO()):
            m.ReportCard(cfg, output_path=out).generate()
        with open(out, "rb") as f:
            return f.read()
    finally:
        os.remove(out)


# The cards both versions have. A card only one side has cannot show "no
# change": master's from_cli_args() falls back to its default card for a slug it
# does not know, so comparing one would report a difference that is only the
# fallback. It is named here instead, and the comparison below is of the cards
# that exist on both sides.
SHARED = [g for g in NEW.GPUS if g in OLD.GPUS]
print(f"cards only in new: {[g for g in NEW.GPUS if g not in OLD.GPUS]}  "
      f"only in {BASE_REF}: {[g for g in OLD.GPUS if g not in NEW.GPUS]}")

n = comp_diffs = story_n = story_diffs = pdf_n = pdf_diffs = cfg_diffs = 0
new_only, old_only, samples = set(), set(), []
for gpu in SHARED:
    for count, model, prec, fp8kv, (ctx, conc), no_nv in itertools.product(COUNTS, MODELS, PRECS, KVS, CTX, NV):
        if no_nv and NEW.GPUS[gpu]["form"] != "sxm":
            continue  # nvlink_for() downgrades either way
        a = args_for(gpu, count, model, prec, fp8kv, ctx, conc, no_nv)
        # A side may refuse a plan outright (FP8 weights on a card vLLM can't run them
        # on, since feat/rocm-guidance). A refusal is an outcome to compare, not a
        # crash: the same on both sides is no difference, a change is one.
        refused = []
        with contextlib.redirect_stdout(io.StringIO()):
            for side in (OLD, NEW):
                try:
                    refused.append((side.from_cli_args(a), None))
                except ValueError as e:
                    refused.append((None, str(e)))
        (ca, ra), (cb, rb) = refused
        if ra or rb:
            n += 1
            if ra != rb:
                cfg_diffs += 1
                if len(samples) < 12: samples.append(f"{gpu} x{count} {model} {prec}: refused {ra!r} -> {rb!r}")
            continue
        n += 1
        for k in set(ca) | set(cb):
            if k == "gpu":
                if ca[k] != cb[k]:
                    cfg_diffs += 1; samples.append(f"{gpu}: cfg gpu row differs")
                continue
            if ca.get(k) != cb.get(k):
                cfg_diffs += 1
                if len(samples) < 12: samples.append(f"{gpu} x{count} {model} {prec}: cfg[{k}] {ca.get(k)!r} -> {cb.get(k)!r}")
        pa, pb = OLD.compute(ca), NEW.compute(cb)
        for k in pa:
            if k not in pb:
                old_only.add(k); continue
            if pa[k] != pb[k] or type(pa[k]) is not type(pb[k]):
                comp_diffs += 1
                if len(samples) < 12: samples.append(f"{gpu} x{count} {model} {prec} kv{fp8kv} {ctx}/{conc} nonv{no_nv}: {k} {pa[k]!r} -> {pb[k]!r}")
        for k in pb:
            if k not in pa: new_only.add(k)
        if pb.get("throughput_modelled") is not True:
            comp_diffs += 1; samples.append(f"{gpu}: throughput_modelled={pb.get('throughput_modelled')!r}")
        if OLD.build_vllm_cmd(ca, pa) != NEW.build_vllm_cmd(cb, pb):
            comp_diffs += 1; samples.append(f"{gpu} x{count} {model} {prec}: command differs")
        if not pa["fits"] and OLD.boards_advice(ca, pa) != NEW.boards_advice(cb, pb):
            comp_diffs += 1; samples.append(f"{gpu} x{count} {model} {prec}: advice differs")
        if n % STRIDE == 0:
            story_n += 1
            sa, sb = story_strings(OLD, ca), story_strings(NEW, cb)
            if sa != sb:
                story_diffs += 1
                if len(samples) < 12:
                    i = next((i for i, (x, y) in enumerate(zip(sa, sb)) if x != y), min(len(sa), len(sb)))
                    samples.append(f"{gpu} x{count} {model} {prec} kv{fp8kv} {ctx}/{conc} nonv{no_nv}: story differs at {i}: old {sa[i:i+2]!r} new {sb[i:i+2]!r} (lengths {len(sa)} vs {len(sb)})")
        if n % PDF_STRIDE == 0:
            pdf_n += 1
            if pdf_bytes(OLD, ca) != pdf_bytes(NEW, cb):
                pdf_diffs += 1
                if len(samples) < 12: samples.append(f"{gpu} x{count} {model} {prec} kv{fp8kv} {ctx}/{conc} nonv{no_nv}: PDF bytes differ")

# from_json (preset and raw branches) and interactive_mode on a few cards
import json, tempfile
jn = jd = 0
for gpu in ("a100-40", "h100-80", "rtx4090-24", "t4-16", "b200-192"):
    for spec in ({"preset": "llama31-70b", "gpu": gpu, "bpp": 1, "ctx": 16384, "conc": 8, "n_gpu": 4, "nvlink": True},
                 {"params": 8, "layers": 32, "kv_heads": 8, "gpu": gpu, "n_gpu": 2, "nvlink": False, "kv_bpp": 1}):
        fd, p = tempfile.mkstemp(suffix=".json"); os.close(fd)
        with open(p, "w") as f: json.dump(spec, f)
        try:
            with contextlib.redirect_stdout(io.StringIO()):
                ca, cb = OLD.from_json(p), NEW.from_json(p)
        finally:
            os.remove(p)
        jn += 1
        if story_strings(OLD, ca) != story_strings(NEW, cb) or OLD.compute(ca) != {k: v for k, v in NEW.compute(cb).items() if k in OLD.compute(ca)}:
            jd += 1; samples.append(f"from_json {gpu} {spec}: differs")
    outs = []
    for m in (OLD, NEW):
        # Each version's own menu number: the two lists need not agree once the
        # catalog grows, and this compares a card, not a menu position.
        answers = [str(list(m.PRESETS).index("llama31-8b") + 1), str(list(m.GPUS).index(gpu) + 1), "2", "y", "1", "n", "8192", "4"]
        ans = list(answers) if m.supports_nvlink(m.GPUS[gpu]) else [a for i, a in enumerate(answers) if i != 3]
        with unittest.mock.patch("builtins.input", side_effect=ans), contextlib.redirect_stdout(io.StringIO()):
            outs.append(m.interactive_mode())
    jn += 1
    a, b = outs
    if {k: v for k, v in a.items() if k != "gpu"} != {k: v for k, v in b.items() if k != "gpu"}:
        jd += 1; samples.append(f"interactive {gpu}: {a} vs {b}")

print(f"from_cli_args cfg: {n} configurations, {cfg_diffs} differences")
print(f"compute() + command + advice: {n} configurations, {comp_diffs} differences")
print(f"  fields only in new: {sorted(new_only)}  only in master: {sorted(old_only)}")
print(f"PDF story strings: {story_n} configurations, {story_diffs} differing")
print(f"PDF bytes (pinned clock, invariant mode): {pdf_n} configurations, {pdf_diffs} differing")
print(f"from_json / interactive_mode: {jn} checks, {jd} differing")
for s in samples:
    print("  " + s)

# A comparison that compared nothing is not a pass. Its JS twin printed
# "0 configurations, 0 differing surfaces" for a whole run and had no exit code
# at all, so nothing could tell that apart from a clean result.
if 0 in (n, pdf_n, jn):
    sys.exit(f"compared nothing (states {n}, pdfs {pdf_n}, json/interactive {jn}) "
             f"— this proves nothing")
